### PR TITLE
cleaned up TypeScript in coders.ts

### PR DESCRIPTION
**Pull request by pcoet**
**Please** add a meaningful description for your change here

This fixes 22 problems caught by eslint in coders.ts. The file should be clean and `any` free now. Let me know if I've misunderstood the intention of any of the types... I'm just looking at this code for the first time. There are probably some `unknown` types that could be narrower.

@robertwb 
@pabloem 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by aaltay** [comment]
R: @damccorm

File: sdks/typescript/src/apache_beam/coders/coders.ts [hunk 1]
```diff
@@ -114,20 +114,31 @@ export interface Coder<T> {
...
-function writeByteCallback(val, buf, pos) {
+function writeByteCallback(
+  val: number,
+  buf: { [x: string]: number },
+  pos: number
+) {
   buf[pos] = val & 0xff;
 }
 
+interface HackedWriter extends Writer {
```
Comment (damccorm) [inline sdks/typescript/src/apache_beam/coders/coders.ts:125]: Should we export this? Presumably we're going to want it in the caller eventually

**Review by damccorm — APPROVED** [review]
Just had 1 small nit - thank you!

File: sdks/typescript/src/apache_beam/coders/coders.ts (hunk 1 above)
Comment (pcoet) [reply sdks/typescript/src/apache_beam/coders/coders.ts]: Yes, done.

**Comment by pcoet** [comment]
This should be ready now. I had to update a few types after resolving a merge conflict.

**Comment by damccorm** [comment]
@pabloem or @robertwb would you mind taking a look/merging?

**Pull request merged**